src/simulation/Benchmark.cpp

```diff
@@ -0,0 +1,402 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "simulation/Benchmark.h"
+
+#include "bucket/BucketManager.h"
+#include "database/Database.h"
+#include "herder/Herder.h"
+#include "ledger/LedgerDelta.h"
+#include "util/Logging.h"
+#include "util/make_unique.h"
+#include <algorithm>
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <random>
+#include <vector>
+
+namespace stellar
+{
+
+const char* Benchmark::LOGGER_ID = "Benchmark";
+
+size_t Benchmark::MAXIMAL_NUMBER_OF_TXS_PER_LEDGER = 1000;
+
+Benchmark::Benchmark(medida::MetricsRegistry& registry, uint32_t txRate,
+                     std::unique_ptr<TxSampler> sampler)
+    : mIsRunning(false)
+    , mTxRate(txRate * LoadGenerator::STEP_MSECS / 1000)
+    , mMetrics(initializeMetrics(registry))
+    , mSampler(std::move(sampler))
+{
+}
+
+Benchmark::~Benchmark()
+{
+    if (mIsRunning)
+    {
+        stopBenchmark();
+    }
+}
+
+void
+Benchmark::startBenchmark(Application& app)
+{
+    if (mIsRunning)
+    {
+        throw std::runtime_error{"Benchmark already started"};
+    }
+    mIsRunning = true;
+    mBenchmarkTimeContext =
+        make_unique<medida::TimerContext>(mMetrics.benchmarkTimer.TimeScope());
+    scheduleLoad(app,
+                 std::chrono::milliseconds{LoadGenerator::STEP_MSECS});
+}
+
+Benchmark::Metrics
+Benchmark::initializeMetrics(medida::MetricsRegistry& registry)
+{
+    return Benchmark::Metrics(registry);
```

Comment:
this looks lime something that is not required to compile on all C++11 compilers, as Benchmark::Metrics does not have copy-constructor and RVO is not required in C++11